### PR TITLE
release: promote beta to main (hermes dashboard-auth fix)

### PR DESCRIPTION
**Pull request by KrasimirKralev**
Second promotion of the day, cut so that a Hermes device can be reset onto a
release build that contains the fix written for it.

## What this adds over `505cab6`

**The Hermes dashboard-auth failure is fixed at its cause.** Provisioning a
Hermes device could report that a freshly written password and hash did not
verify, while both were correct. The cause was a lost update: the installer
starts the web service — which asynchronously rewrites the Hermes config — and
then, a few steps later, writes the dashboard block and re-reads it to check.
A write landing between those two points removed the block, and the check
reported it as a credential mismatch.

The same race had a second, quieter phase: when the competing write landed
after the check instead of before it, provisioning reported success on a device
that had no auth provider at all.

This change closes the race, separates "the check could not run" from "the
credentials disagree" so the message names the real condition, and makes a
failed provisioning step visible in the summary rather than reported as
success.

## Why now

Devices are being prepared for shipping. A Hermes unit reset onto `main`
without this would carry a provisioning step that can fail confusingly, or
succeed misleadingly — neither is something to hand to a customer.

## Verification

Beyond CI: the failure was reproduced deterministically before the fix, and the
OpenClaw devices flashed from `505cab6` earlier today are unaffected by it, so
this narrows rather than widens what is already in the field.

## Summary by CodeRabbit

- **Bug Fixes**
  - Provisioning and service-validation failures now remain visible in the final installation summary.
  - Installations correctly return a non-zero exit status when setup is incomplete.
  - Hermes dashboard authentication checks now detect missing, invalid, or unusable credentials.
  - Unauthorized proxy responses are no longer reported as healthy.

- **Improvements**
  - Installation output includes failed steps, remediation commands, and a machine-readable status (`STATUS=ok` or `STATUS=incomplete`).
  - Concurrent Hermes configuration updates are coordinated to prevent conflicting changes.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -61,6 +61,37 @@ PROJECT_DIR="/home/clawbox/clawbox"
...
+# reach the operator's summary AND the caller's exit status, or a flash host
+# prints "Setup: 1/1 succeeded" over an install that told itself it was broken.
+# So every non-fatal failure is recorded here, the summary lists them, install.sh
+# exits non-zero, and a machine-readable marker is left for the flash host.
+PROVISION_FAILURES=()
+PROVISION_STATUS_FILE="${CLAWBOX_PROVISION_STATUS_FILE:-/etc/clawbox/provision-status}"
+
+record_provision_failure() {
+  PROVISION_FAILURES+=("$1")
+}
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply install.sh:77]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: install.sh [hunk 2]
```diff
@@ -61,6 +61,37 @@ PROJECT_DIR="/home/clawbox/clawbox"
...
+  dir="$(dirname "$PROVISION_STATUS_FILE")"
+  mkdir -p "$dir" 2>/dev/null || true
+  {
+    echo "# Written by install.sh at the end of a full install. Machine-readable."
+    echo "STATUS=$status"
+    echo "FAILED_STEPS=$*"
+    echo "TIMESTAMP=$(date -u +%Y-%m-%dT%H:%M:%SZ 2>/dev/null || true)"
+  } > "$PROVISION_STATUS_FILE" 2>/dev/null || true
+  chmod 644 "$PROVISION_STATUS_FILE" 2>/dev/null || true
+}
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply install.sh:93]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: scripts/setup-hermes-dashboard-auth.sh [hunk 3]
```diff
@@ -195,25 +276,131 @@ except Exception as exc:
...
+  # config.yaml now holds the session-signing secret and the password hash.
+  chmod 600 "$HERMES_CONFIG"
+  for bak in "$HERMES_CONFIG".bak*; do
+    [ -f "$bak" ] && chmod 600 "$bak" 2>/dev/null || true
+  done
+
+  # Store the plaintext password for the proxy ONLY (clawbox-owned, 0600).
+  ( umask 077; printf '%s' "$pw" > "$PWFILE" )
+  chmod 600 "$PWFILE"
+  return 0
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply scripts/setup-hermes-dashboard-auth.sh:292]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: scripts/setup-hermes-dashboard-auth.sh [hunk 4]
```diff
@@ -195,25 +276,131 @@ except Exception as exc:
...
+      log "ERROR: the password we just wrote does not verify against the password_hash we just wrote. This is a hashing fault in this script, NOT an environment problem and NOT a lost write." >&2
+      exit 2
+      ;;
+    "$CREDS_NOT_CONFIGURED"|"$CREDS_PW_MISSING")
+      if [ "$attempt" -lt "$max_attempts" ]; then
+        log "WARNING: the dashboard auth block / password file we just wrote to $HERMES_CONFIG is already gone — a concurrent process rewrote the file. The credentials generated were correct; the write was lost. Retrying under $CONFIG_LOCK (attempt $attempt/$max_attempts)." >&2
+        continue
+      fi
+      log "ERROR: after $attempt attempts our dashboard auth block did not survive in $HERMES_CONFIG — another process keeps rewriting the file and is not honouring $CONFIG_LOCK. This is a write/serialisation fault, NOT a credential mismatch: the password and hash were correct on every attempt." >&2
+      exit 3
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply scripts/setup-hermes-dashboard-auth.sh:391]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: src/tests/unit/hermes-config-lock.test.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  it("both take the lock before they touch config.yaml", () => {
+    // The auth script acquires before its mint; the registrar acquires before
+    // its PyYAML reconcile AND holds it across the `hermes tools disable` CLI
+    // call (which does its own wide load->save_config on the same file).
+    expect(AUTH_SRC).toMatch(/acquire_config_lock[\s\S]*mint_credentials/);
+    const regTail = REGISTER_SRC.slice(REGISTER_SRC.indexOf("acquire_config_lock\n\nexport"));
+    expect(regTail).toContain("acquire_config_lock");
+    expect(REGISTER_SRC.indexOf("acquire_config_lock\n\nexport")).toBeLessThan(
+      REGISTER_SRC.indexOf("tools disable browser"),
+    );
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/hermes-config-lock.test.ts:50]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: src/tests/unit/hermes-config-lock.test.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      start=$(date +%s.%N)
+      CLAWBOX_ROOT="${root}" HERMES_CONFIG="${configPath}" bash "${AUTH}" >/dev/null 2>&1
+      end=$(date +%s.%N)
+      wait $hold
+      awk -v s="$start" -v e="$end" 'BEGIN{printf "%.3f", e - s}'
+    `;
+    const proc = spawnSync("bash", ["-c", script], { encoding: "utf-8", timeout: 20000 });
+    const elapsed = parseFloat(proc.stdout.trim() || "0");
+    // Held ~0.8s, started ~0.15s in, so the auth script should wait ≳0.5s.
+    expect(elapsed).toBeGreaterThan(0.5);
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/hermes-config-lock.test.ts:111]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: src/tests/unit/hermes-dashboard-auth-yaml.test.ts [hunk 7]
```diff
@@ -34,6 +34,82 @@ function run(root: string, configPath: string, username?: string) {
...
+        "pw=sys.argv[1].encode(); salt=bytes(range(16))",
+        "dk=hashlib.scrypt(pw,salt=salt,n=2**14,r=8,p=1,dklen=32,maxmem=0)",
+        "print('scrypt$%d$%d$%d$%s$%s'%(2**14,8,1,base64.b64encode(salt).decode(),base64.b64encode(dk).decode()))",
+      ].join("\n"),
+      plaintext,
+    ],
+    { encoding: "utf-8" },
+  );
+  return proc.stdout.trim();
+}
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/hermes-dashboard-auth-yaml.test.ts:93]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: src/tests/unit/hermes-dashboard-auth-yaml.test.ts [hunk 8]
```diff
@@ -146,3 +222,103 @@ describe.runIf(RUNNABLE)("hermes dashboard auth block", () => {
...
+/** Full provision with extra env (used to inject a broken PATH). */
+function run2(root: string, configPath: string, extraEnv: NodeJS.ProcessEnv) {
+  const env: NodeJS.ProcessEnv = {
+    ...process.env,
+    CLAWBOX_ROOT: root,
+    HERMES_CONFIG: configPath,
+    ...extraEnv,
+  };
+  return spawnSync("bash", [SCRIPT], { encoding: "utf-8", env });
+}
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/hermes-dashboard-auth-yaml.test.ts:324]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.

File: src/tests/unit/install-hermes-edition-step.test.ts [hunk 9]
```diff
@@ -110,6 +110,93 @@ describe("the TS and shell harness predicates agree", () => {
...
+
+  it("no longer counts a proxy 401 as healthy on hermes", () => {
+    // 401 is the desynced-SSO symptom (see hermes-dashboard-proxy.js); a healthy
+    // browserless probe gets a 302/403, so a 401 must fail the check. Assert the
+    // case pattern precisely — the word "401" still appears in the DESYNCED
+    // message, and that is correct.
+    const fn = extractShellFunction("step_validate_services");
+    expect(fn).toContain("2*|3*|403) ;;");
+    expect(fn).not.toContain("2*|3*|401|403");
+  });
```
[the remark this answers is not in the document]
Comment (KrasimirKralev) [reply src/tests/unit/install-hermes-edition-step.test.ts:191]: Acknowledged, not dismissed — deferring to a follow-up branch so a time-critical device promotion is not held up.

The race this PR fixes is the shipping-relevant part and is already in. The findings here are edge cases on top of it (unwritable marker path, torn read during a concurrent write, immediate retry, test assertions that can degrade silently), none of which are regressions introduced by this promotion.

They are being fixed on `fix/provision-status-integrity`, with the marker-integrity finding treated as the priority: a run that cannot write its verdict must not be readable as the previous run’s success. Tracking so this is not lost.
